core/domain/state_domain.py

```diff
@@ -353,192 +406,1900 @@ def __init__(
+                interaction_dict['customization_args']
+            )
+        )
+
+        return cls(
+            interaction_dict['id'],
+            customization_args,
+            [AnswerGroup.from_dict(h)
+             for h in interaction_dict['answer_groups']],
+            default_outcome_dict,
+            interaction_dict['confirmed_unclassified_answers'],
+            [Hint.from_dict(h) for h in interaction_dict['hints']],
+            solution_dict)
+
+    def to_proto(self):
+        """Returns a proto representation of the interaction object.
+
+        Returns:
+            InteractionInstance. The InteractionInstance proto object.
+        """
+        interaction_proto = None
+        if self.id == 'Continue':
+            interaction_proto = state_pb2.InteractionInstance(
+                continue_instance=self._to_continue_proto(self))
+
+        if self.id == 'FractionInput':
+            interaction_proto = state_pb2.InteractionInstance(
+                fraction_input=self._to_fraction_interaction_proto(
+                    self))
+
+        if self.id == 'ItemSelectionInput':
+            interaction_proto = state_pb2.InteractionInstance(
+                item_selection_input=(
+                    self._to_item_selection_single_interaction_proto(
+                        self)))
+
+        if self.id == 'MultipleChoiceInput':
+            interaction_proto = state_pb2.InteractionInstance(
+                multiple_choice_input=(
+                    self._to_multiple_choice_interaction_proto(
+                        self)))
+
+        if self.id == 'NumericInput':
+            interaction_proto = state_pb2.InteractionInstance(
+                numeric_input=self._to_numeric_interaction_proto(
+                    self))
+
+        if self.id == 'TextInput':
+            interaction_proto = state_pb2.InteractionInstance(
+                text_input=self._to_text_input_interaction_proto(
+                    self))
+
+        if self.id == 'RatioExpressionInput':
+            interaction_proto = state_pb2.InteractionInstance(
+                ratio_expression_input=(
+                    self._to_ratio_expression_interaction_proto(
+                        self)))
+
+        if self.id == 'ImageClickInput':
+            interaction_proto = state_pb2.InteractionInstance(
+                image_click_input=(
+                    self._to_image_click_interaction_proto(
+                        self)))
+
+        if self.id == 'DragAndDropSortInput':
+            interaction_proto = state_pb2.InteractionInstance(
+                drag_and_drop_sort_input=(
+                    self._to_drag_and_drop_interaction_proto(
+                        self)))
+
+        if self.id == 'EndExploration':
+            interaction_proto = state_pb2.InteractionInstance(
+                end_exploration=self._to_end_exploration_proto())
+
+        return interaction_proto
+
+    @classmethod
+    def _to_continue_proto(cls, interaction):
```

Comment:
Can you keep the interaction-specific stuff in the relevant places in extensions/interactions, and retrieve it through interaction_registry() instead?

Each interaction should have a single to_customization_args_proto() method that takes a set of customization args and makes the proto, and similarly for to_solution_proto(). (Some of these may raise NotImplementedError or provide an empty proto if they are irrelevant.) Then you can have a generic function here that assembles the proto based on the interaction ID. The outcome proto and other things that are not interaction-specific can be handled here.

Reply:
with the `to_customization_args_proto` and `to_solution_proto` function, I also have to shift the per interaction related rule_specs and answer groups functions.